# Python and PANDAS

## Requires Winnipeg weather dataset (winnipeg.csv attached to this worksheet)
#### can login sign up and get weather data from https://www.visualcrossing.com/ for learning purposes or use the given dataset

### REMEMBER TO REPLACE THE FILE PATH!

In [2]:
import pandas as pd

In [5]:
data = pd.read_csv(r"C:\Users\kyled\Downloads\Assignment 2 Part 1 Starter Files(1)\winnipeg.csv")

In [6]:
data

name    datetime  tempmax  tempmin  temp  feelslikemax  feelslikemin  \
0   winnipeg  2023-12-21      1.1     -3.0  -0.4          -2.5          -6.9   
1   winnipeg  2023-12-22     -2.3     -5.3  -3.8          -4.7          -9.0   
2   winnipeg  2023-12-23      0.8     -3.4  -1.6          -4.3          -9.0   
3   winnipeg  2023-12-24     -1.7     -7.0  -3.1          -6.9         -13.6   
4   winnipeg  2023-12-25     -4.0     -9.6  -7.3          -9.6         -15.4   
5   winnipeg  2023-12-26     -3.6     -9.1  -6.9          -8.5         -14.9   
6   winnipeg  2023-12-27     -2.8     -8.1  -5.8          -7.6         -13.5   
7   winnipeg  2023-12-28     -3.0     -6.7  -5.3          -8.3         -12.8   
8   winnipeg  2023-12-29     -5.8    -14.4 -10.9         -11.4         -21.6   
9   winnipeg  2023-12-30    -10.0    -15.3 -13.6         -17.1         -22.8   
10  winnipeg  2023-12-31     -8.6    -14.5 -12.0         -14.4         -21.4   
11  winnipeg  2024-01-01    -10.9    -14.2 -12.6         -17.8         -21.4   
12  winnipeg  2024-01-02    -13.2    -17.0 -14.8         -22.0         -27.5   
13  winnipeg  2024-01-03    -16.3    -18.9 -17.6         -23.8         -27.6   
14  winnipeg  2024-01-04    -16.3    -21.7 -19.8         -25.3         -31.8   

    feelslike   dew  humidity  ...  solarenergy  uvindex  severerisk  \
0        -4.1  -3.5      79.9  ...          2.3        2          10   
1        -6.9  -4.9      92.2  ...          2.2        1          10   
2        -7.0  -2.9      91.1  ...          3.0        1          10   
3        -9.2  -5.3      84.5  ...          3.8        2          10   
4       -12.9  -9.8      82.5  ...          5.0        2          10   
5       -12.4  -9.9      79.5  ...          4.9        2          10   
6       -11.0  -8.5      81.3  ...          3.5        2          10   
7       -10.8  -7.5      84.3  ...          4.8        2          10   
8       -17.9 -13.6      81.0  ...          3.1        2          10   
9       -20.9 -17.3      74.2  ...          5.2        3          10   
10      -18.2 -14.3      83.1  ...          4.8        2          10   
11      -19.7 -15.0      82.0  ...          4.5        2          10   
12      -24.9 -17.7      78.6  ...          3.2        2          10   
13      -26.1 -21.2      73.4  ...          4.9        3          10   
14      -29.2 -23.7      71.2  ...          5.1        3          10   

                sunrise               sunset  moonphase        conditions  \
0   2023-12-21T08:24:06  2023-12-21T16:29:16       0.31  Partially cloudy   
1   2023-12-22T08:24:36  2023-12-22T16:29:47       0.35  Partially cloudy   
2   2023-12-23T08:25:03  2023-12-23T16:30:21       0.38  Partially cloudy   
3   2023-12-24T08:25:27  2023-12-24T16:30:57       0.42  Partially cloudy   
4   2023-12-25T08:25:48  2023-12-25T16:31:37       0.45  Partially cloudy   
5   2023-12-26T08:26:05  2023-12-26T16:32:20       0.50  Partially cloudy   
6   2023-12-27T08:26:20  2023-12-27T16:33:05       0.52  Partially cloudy   
7   2023-12-28T08:26:31  2023-12-28T16:33:53       0.56  Partially cloudy   
8   2023-12-29T08:26:40  2023-12-29T16:34:44       0.59  Partially cloudy   
9   2023-12-30T08:26:45  2023-12-30T16:35:38       0.62             Clear   
10  2023-12-31T08:26:47  2023-12-31T16:36:34       0.65  Partially cloudy   
11  2024-01-01T08:26:46  2024-01-01T16:37:33       0.69  Partially cloudy   
12  2024-01-02T08:26:41  2024-01-02T16:38:35       0.72          Overcast   
13  2024-01-03T08:26:34  2024-01-03T16:39:39       0.75  Partially cloudy   
14  2024-01-04T08:26:23  2024-01-04T16:40:45       0.78  Partially cloudy   

                             description               icon  \
0      Partly cloudy throughout the day.  partly-cloudy-day   
1      Partly cloudy throughout the day.  partly-cloudy-day   
2      Partly cloudy throughout the day.  partly-cloudy-day   
3      Partly cloudy throughout the day.  partly-cloudy-day   
4      Partly cloudy throu

# How to Analyze DataFrames

# .head()  
### It shows the first N rows in the data(by default, N=5)


In [7]:
data.head(n=5)

name    datetime  tempmax  tempmin  temp  feelslikemax  feelslikemin  \
0  winnipeg  2023-12-21      1.1     -3.0  -0.4          -2.5          -6.9   
1  winnipeg  2023-12-22     -2.3     -5.3  -3.8          -4.7          -9.0   
2  winnipeg  2023-12-23      0.8     -3.4  -1.6          -4.3          -9.0   
3  winnipeg  2023-12-24     -1.7     -7.0  -3.1          -6.9         -13.6   
4  winnipeg  2023-12-25     -4.0     -9.6  -7.3          -9.6         -15.4   

   feelslike  dew  humidity  ...  solarenergy  uvindex  severerisk  \
0       -4.1 -3.5      79.9  ...          2.3        2          10   
1       -6.9 -4.9      92.2  ...          2.2        1          10   
2       -7.0 -2.9      91.1  ...          3.0        1          10   
3       -9.2 -5.3      84.5  ...          3.8        2          10   
4      -12.9 -9.8      82.5  ...          5.0        2          10   

               sunrise               sunset  moonphase        conditions  \
0  2023-12-21T08:24:06  2023-12-21T16:29:16       0.31  Partially cloudy   
1  2023-12-22T08:24:36  2023-12-22T16:29:47       0.35  Partially cloudy   
2  2023-12-23T08:25:03  2023-12-23T16:30:21       0.38  Partially cloudy   
3  2023-12-24T08:25:27  2023-12-24T16:30:57       0.42  Partially cloudy   
4  2023-12-25T08:25:48  2023-12-25T16:31:37       0.45  Partially cloudy   

                         description               icon              stations  
0  Partly cloudy throughout the day.  partly-cloudy-day  CXWN,CYWG,C8306,CWNK  
1  Partly cloudy throughout the day.  partly-cloudy-day                   NaN  
2  Partly cloudy throughout the day.  partly-cloudy-day                   NaN  
3  Partly cloudy throughout the day.  partly-cloudy-day                   NaN  
4  Partly cloudy throughout the day.  partly-cloudy-day                   NaN  

[5 rows x 33 columns]

# .shape
### It shows the total no. of rows and no. of columns of the dataframe

In [13]:
# Shows numbers of rows and columns, cannot use () as there is nothing to call.
data.shape

(15, 33)

# .index 
### provides the index of the dataframe

In [15]:
# Shows the idnex of the dataframe, meaning that it displays the range of integers that are used for indexing rows.
data.index

RangeIndex(start=0, stop=15, step=1)

# .dtypes
### It shows the data-type of each column


In [16]:
# Gives the datatypes of each column/
data.dtypes

name                 object
datetime             object
tempmax             float64
tempmin             float64
temp                float64
feelslikemax        float64
feelslikemin        float64
feelslike           float64
dew                 float64
humidity            float64
precip              float64
precipprob          float64
precipcover         float64
preciptype           object
snow                  int64
snowdepth           float64
windgust            float64
windspeed           float64
winddir             float64
sealevelpressure    float64
cloudcover          float64
visibility          float64
solarradiation      float64
solarenergy         float64
uvindex               int64
severerisk            int64
sunrise              object
sunset               object
moonphase           float64
conditions           object
description          object
icon                 object
stations             object
dtype: object

# .columns
### It shows the name of each column

In [19]:
# Gives the name of each column.
data.columns

Index(['name', 'datetime', 'tempmax', 'tempmin', 'temp', 'feelslikemax',
       'feelslikemin', 'feelslike', 'dew', 'humidity', 'precip', 'precipprob',
       'precipcover', 'preciptype', 'snow', 'snowdepth', 'windgust',
       'windspeed', 'winddir', 'sealevelpressure', 'cloudcover', 'visibility',
       'solarradiation', 'solarenergy', 'uvindex', 'severerisk', 'sunrise',
       'sunset', 'moonphase', 'conditions', 'description', 'icon', 'stations'],
      dtype='object')

# .unique()
### Shows all unique values in a column. It can be applied to a single column only, not the whole dataframe.
### eg. dataname['Column name'].unique()

# .nunique()
### shows the unique values in each column. Can be applied to both single columns and the whole dataframe.


In [22]:
# This displays unique values in the snow column, and returns a numpy array containing all the unique values within the series.
# Cannot include anything within the bracket.
data['snow'].unique()

array([0], dtype=int64)

In [23]:
# This instead finds the number of unique values in the SolarEnergy column.
# This method returns an integer that represents the count of unique values within the series.
data['solarenergy'].nunique()

13

# .count()
### Shows the total number of non-null values in each column.


# .value_counts()
### In a column, it shows all the unique values with their count. Can be applied to a single column only.


In [30]:
# Show sthe number of non-nnull values in each column.
# Can use without giving a column name, to see every column's count.
# data.count()
data['dew'].count()

15

In [26]:
# Shows the unique valuess 
data['dew'].value_counts()

dew
-3.5     1
-4.9     1
-2.9     1
-5.3     1
-9.8     1
-9.9     1
-8.5     1
-7.5     1
-13.6    1
-17.3    1
-14.3    1
-15.0    1
-17.7    1
-21.2    1
-23.7    1
Name: count, dtype: int64

# .info()
### Provides basic information about the dataframe

# 1. Find the unique values from column 'temp'.

# 2. Find the number of times when condition in Partially cloudy

# 3. Display the data when the tempmax was exactuly -2.3

# 4. Find out all Null Values in the stations(can use isnull() function and sum() function to find the total)

# .mean()
## 5. Calculate mean value of solarenergy column. 

# .std()
## 6. Calculate Standard Deviation of humidity column.

# .var()
## 7. Calculate the variance of "dew" column

# .max() .min()
## Find the maximum and minimum temperate from temp column